damnit/api.py

```diff
@@ -0,0 +1,279 @@
+            elif dset.ndim == 0:
+                # Scalars
+                return dset[()]
+            else:
+                # Otherwise, return a Numpy array
+                return group["data"][()]
+
+    def summary(self):
+        result = self._db.conn.execute("""
+            SELECT value, max(version) FROM run_variables
+            WHERE proposal=? AND run=? AND name=?
+        """, (self.proposal, self.run, self._name)).fetchone()
+
+        if result is None:
+            # This should never be reached unless the variable is deleted
+            # after creating the VariableData object.
+            raise RuntimeError(f"Could not find value for '{self._name}' in p{self.proposal}, r{self._name}")
+        else:
+            return result[0]
+
+    def __repr__(self):
+        return f"<VariableData for '{self.name}' in p{self.proposal}, r{self.run}>"
+
+
+class RunVariables:
+    def __init__(self, db_dir, run):
+        self._db = DamnitDB.from_dir(db_dir)
+        self._proposal = self._db.metameta["proposal"]
+        self._run = run
+        self._data_format_version = self._db.metameta["data_format_version"]
+        self._h5_path = Path(db_dir) / f"extracted_data/p{self._proposal}_r{self._run}.h5"
+
+    @property
+    def proposal(self):
+        return self._proposal
+
+    @property
+    def run(self):
+        return self._run
+
+    def __getitem__(self, name):
+        key_locs = self._key_locations()
+        titles = self._var_titles()
+
+        if name not in key_locs and name not in titles:
+            raise KeyError(f"Variable data for '{name}' not found for p{self.proposal}, r{self.run}")
+
+        if name in titles:
+            name = titles[name]
+
+        return VariableData(name, self.proposal, self.run,
+                            self._h5_path, self._data_format_version,
+                            self._db, key_locs[name])
+
+    def _key_locations(self):
+        # Read keys from the HDF5 file
+        with h5py.File(self.file) as f:
+            all_keys = { name: False for name in f.keys() }
+        del all_keys[".reduced"]
+
+        # And the keys from the database
+        user_vars = list(self._db.get_user_variables().keys())
+        user_vars.append("comment")
+
+        for var_name in user_vars:
+            result = self._db.conn.execute("""
+                SELECT name, value, max(version) FROM run_variables
+                WHERE proposal=? AND run=? AND name=?
+            """, (self.proposal, self.run, var_name)).fetchone()
+            if result is not None and result[1] is not None:
+                all_keys[var_name] = True
+
+        return all_keys
+
+    def keys(self):
+        return sorted(self._key_locations().keys())
+
+    def _var_titles(self):
+        result = self._db.conn.execute("SELECT name, title FROM variables").fetchall()
+        return { row[1] if row[1] is not None else row[0]: row[0] for row in result
+                 if row[0] in self.keys() }
```

Comment:
I think this calls `self.keys()` again for each key that it wants to check. That might be fast enough that it doesn't really matter, but...

```suggestion
        keys = self._key_locations()
        return { row[1] if row[1] is not None else row[0]: row[0] for row in result
                 if row[0] in keys }
```